### PR TITLE
[dynamo] Leave DTensor structural values alone when pruning a guarded object's attributes

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.17.0) (oldest at bottom):
* #197761
* #197760
* #197759
* #197758
* __->__ #197821
* #197835
* #198034
* #198026

**Problem.** Attribute pruning is keyed by id, so an unguarded DTensor structural object on a guarded holder (a `Shard` placement or a `DeviceMesh` on a config object or a module) could be the very object a `DTensorSpec` elsewhere in the state rebuilds itself from, and substituting it would make that spec's guard miss forever.

**Fix.** `_prune_unguarded_attributes` never registers a value whose type lives under `torch.distributed.tensor` or `torch.distributed.device_mesh` (`_DTENSOR_MODULES`).

**Summary.** The previous commit excludes torch-typed RECEIVERS from pruning because their fields are structural; this commit adds the one class of torch-typed attribute VALUE that must not be pruned either, for the same reason seen from the other side: the value may be shared with a spec that is pickled whole and rebuilds itself from it, and since registration is global, pruning it on the holder substitutes it inside the spec too. The exemption is deliberately this narrow. Every other torch-typed bystander stays prunable, because that is what pruning exists for: a `GradScaler` whose `__getstate__` asserts mid-iteration, or a `DataLoader` iterator whose `__getstate__` refuses, would otherwise take the frame down. The type's module is read defensively, since `__module__` need not be a string, and matched by prefix so that `torch.distributed.tensor.placement_types` and `torch.distributed.device_mesh` are both covered.

## Before and after

```python
shard = Shard(0)
h = _HolderWithGenerator()
h.placement, h.scaler = shard, torch.amp.GradScaler()   # neither is guarded
GuardsStatePickler({id(h): h, id(h.cfg): h.cfg}, {}, {}, {}, buf).dump({"h": h, "shared": shard})
out = load_guards_state(buf.getvalue())
# parent: out["h"].placement and out["shared"] are both the sentinel (one object, pruned by id)
out["h"].placement is out["shared"] and out["shared"] == shard    # now
isinstance(out["h"].scaler, _Missing)                             # before and now
```

`test_an_unguarded_dtensor_placement_is_not_pruned` fails on the parent and passes here; it also pins that the `GradScaler` bystander next to the placement is still pruned.

Test Plan:

```
python test/dynamo/test_guard_serialization.py -k dtensor_placement
python test/dynamo/test_guard_serialization.py
```

Authored with an AI assistant.

Co-Authored-By: Claude Fable 5.1 <noreply@anthropic.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
Thanks for the review. Disposition against head `ed510e43c523`:

> the check reads only `type(attr).__module__`, so it misses user-defined `Placement` subclasses ... the prefix `"torch.distributed.tensor"` exempts far more than DTensor structural values.

**Fixed** in ed510e43c523: the exemption is `isinstance(attr, _dtensor_structural_types())`, a cached tuple of `Placement` (the whole hierarchy, user subclasses included), `DeviceMesh`, `_MeshLayout`, `DTensorSpec` and `TensorMeta`, following the `EQUALS_MATCH` type list in the same file; a `CommDebugMode` or anything else that merely lives under `torch.distributed.tensor` is no longer exempt.

> the new test does not exercise the path the PR is motivated by.

**Fixed**: `test_an_unguarded_placement_subclass_on_a_module_is_not_pruned` puts a user `Partial` subclass on an `nn.Module` holder, where registration is by id, shares it with a second local, and asserts the loaded objects are one real placement while the `GradScaler` next to it is pruned; it fails without the exemption. The user-holder test stays as the receiver-scoped counterpart.

> the "Before and after" block in the PR description is inaccurate

**Fixed** in the commit message (the parent line now says the holder's attribute is the sentinel while the shared object stays real, and the module case is described); the description on GitHub will be refreshed by hand.

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
Thanks for the re-review. Disposition against head `3f747f02d772`:

> the new module test does not distinguish the two behaviours ... confirm it fails without the exemption, and make the value assertion the one that does the pinning.

**Fixed** in 3f747f02d772: the test asserts `out["m"].placement == placement` and the shared object's type first, then identity, then the pruned scaler, with a comment saying why the value assertions carry it. I confirmed again by deleting the exemption locally: the test fails on the equality (the module attribute loads as the sentinel), and passes with the exemption in place.

> `DeviceMesh` now travels whole ... worth a sentence in the commit message

**Fixed**: the commit message weighs the cost (the `_rank_map` tensor and root mesh ride along on a module that stashes an unguarded mesh) against the silent forever-miss.

> A `tuple` of placements ... no change needed.

**Open**: agreed, noted.

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
Thanks. Disposition against head `d7cc0b32217c`:

> an exempted `Placement` subclass loses its Python state on round-trip ... Please add a field to it and assert the field survives; I expect that assertion to fail

**Declined**, with the check you asked for: a `Partial` subclass that sets `self.factor` in `__init__` loses `factor` under plain `pickle.loads(pickle.dumps(p))`, with no guards-state pickler involved, because the pybind `__getstate__` returns `{"reduce_op": ...}`. That is DTensor's own pickling contract, and it already applies to every such placement inside a `DTensorSpec` pickled whole today. The exemption does not change what a placement looks like after pickling; it only stops the module path from turning the spec's copy into the sentinel. The commit message now says so, so the limit is recorded rather than implied away.

> `_FlatLayout` is not in the tuple.

**Declined**: deliberate, as you note; it only appears nested inside a `_MeshLayout`, which travels whole, and the tuple mirrors the `EQUALS_MATCH` list.

> The module test depends on dict insertion order.

**Fixed** in d7cc0b32217c: the test runs both pickle orders and labels its assertions with the order.

> The `DeviceMesh` cost is still undocumented.

**Fixed** last round in the commit message; the GitHub description is refreshed by hand and still shows the older text.

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
Superseded by #198043: the stack was re-created with fresh PR threads (ghstack unlink + resubmit) so reviews start from the current commit and its description rather than the accumulated thread. Same commit content.
